UefiPayloadPkg/SPI/fch_spi_ctrl.c

```diff
@@ -0,0 +1,279 @@
+
+	dump_state("Transaction finished", 1);
+
+	return 0;
+}
+
+VOID spi_init(VOID)
+{
+	DEBUG((EFI_D_INFO, "%a\n", __FUNCTION__));
+	DEBUG((EFI_D_INFO, "%a: %s: SPI BAR at 0x%08lx\n", __FUNCTION__, __func__, spi_get_bar()));
+}
+
+int spi_ctrlr_xfer(CONST struct spi_slave *slave, CONST VOID *dout,
+			__SIZE_TYPE__ bytesout, VOID *din, __SIZE_TYPE__ bytesin)
+{
+	__SIZE_TYPE__ count;
+	UINT8 cmd;
+	UINT8 *bufin = din;
+	CONST UINT8 *bufout = dout;
+
+	if (CONFIG(SOC_AMD_COMMON_BLOCK_SPI_DEBUG))
+		DEBUG((EFI_D_INFO, "%a(%zx, %zx)\n", __FUNCTION__, bytesout,
+			bytesin));
+
+	/* First byte is cmd which cannot be sent through FIFO */
+	cmd = bufout[0];
+	bufout++;
+	bytesout--;
+
+	/*
+	 * Check if this is a write command attempting to transfer more bytes
+	 * than the controller can handle.  Iterations for writes are not
+	 * supported here because each SPI write command needs to be preceded
+	 * and followed by other SPI commands.
+	 */
+	if (bytesout + bytesin > SPI_FIFO_DEPTH) {
+		DEBUG((EFI_D_INFO,
+      "%a: FCH_SC: Too much to transfer, code error!\n", __FUNCTION__));
+		return -1;
+	}
+
+	if (wait_for_ready())
+		return -1;
+
+	spi_write8(SPI_CMD_CODE, cmd);
+	spi_write8(SPI_TX_BYTE_COUNT, bytesout);
+	spi_write8(SPI_RX_BYTE_COUNT, bytesin);
+
+	for (count = 0; count < bytesout; count++)
+		spi_write8(SPI_FIFO + count, bufout[count]);
+
+	if (execute_command())
+		return -1;
+
+	for (count = 0; count < bytesin; count++)
+		bufin[count] = spi_read8(SPI_FIFO + count + bytesout);
+
+	return 0;
+}
+
+int xfer_vectors(CONST struct spi_slave *slave,
+			struct spi_op vectors[], __SIZE_TYPE__ count)
+{
+	return spi_flash_vector_helper(slave, vectors, count, spi_ctrlr_xfer);
+}
+
+int protect_a_range(UINT32 value)
+{
+	UINT32 reg32;
+	UINT8 n;
+
+	/* find a free protection register */
+	for (n = 0; n < MAX_ROM_PROTECT_RANGES; n++) {
+		reg32 = pci_read_config32((struct device *)((VOID *)SOC_LPC_DEV), ROM_PROTECT_RANGE_REG(n));
+		if (!reg32)
+			break;
+	}
+	if (n == MAX_ROM_PROTECT_RANGES)
+		return -1; /* no free range */
+
+	pci_write_config32((struct device *)((VOID *)SOC_LPC_DEV), ROM_PROTECT_RANGE_REG(n), value);
+	return 0;
+}
+
+/*
+ * Protect range of SPI flash defined by region using the SPI flash controller.
+ *
+ * Note: Up to 4 ranges can be protected, though if a particular region requires more than one
+ * range, total number of regions decreases accordingly. Each range can be programmed to 4KiB or
+ * 64KiB granularity.
+ *
+ * Warning: If more than 1 region needs protection, and they need mixed protections (read/write)
+ * than start with the region that requires the most protection. After the restricted commands
+ * have been written, they can't be changed (write once). So if first region is write protection
+ * and second region is read protection, it's best to define first region as read and write
+ * protection.
+ */
+int fch_spi_flash_protect(CONST struct spi_flash *flash, CONST struct region *region,
```

Comment:
We don't need this function. You may remove all ocurrences in source files, structures and headers

Reply:
Removing it